File: stayhub/hasura/metadata.py
"""StayHub's Hasura metadata, as data.

Hasura's own CLI keeps metadata as a tree of YAML files. This is one Python module instead, for
two reasons: it needs no extra binary installed, and the permission rules — which are the
interesting part — can be written once and reused across roles instead of copy-pasted per table.

Apply it with:  python -m scripts.apply     (from the hasura/ directory)

The four roles:

  anonymous  a visitor with no token. Sees PUBLISHED listings and nothing else.
  customer   a signed-in guest. Adds: their own user row, their own bookings.
  host       a guest who also hosts. Adds: their own listings in ANY status, and the
             bookings other people have made at them.
  staff      StayHub staff. Sees everything.

⚠️ The staff role is called `staff`, NOT `admin`. `admin` is RESERVED in Hasura: it is the role
the admin secret grants, it always has unrestricted access, and declaring any permission for it is
rejected with "cannot define permission for admin role" — an error that names the role and not the
fact that the role is special. A JWT can never carry the built-in admin role either, which is the
point: full access requires the secret, not a token.

⚠️ Every role is additive in what a person can *do*, but Hasura roles are NOT hierarchical — a
`host` does not inherit `customer`'s permissions. Each role's rules are declared in full. That is
verbose and it is also why a permission can never be granted by accident through inheritance.
"""

SOURCE = "default"

# Session variables come from the JWT claims FastAPI signs (see app/core/security.py).
# `X-Hasura-User-Id` is the user's public UUID.
SESSION_USER = "X-Hasura-User-Id"

# --- column sets ------------------------------------------------------------------

# ⚠️ NEVER include `password_hash`. A select permission is a column allowlist, and anything listed
# is readable by that role over GraphQL. This is the single most dangerous line in the file.
USER_PUBLIC_COLUMNS = [
    "public_id", "first_name", "last_name", "avatar_url", "host_bio", "is_host", "created_at",
]
USER_SELF_COLUMNS = USER_PUBLIC_COLUMNS + ["email", "role"]

PROPERTY_PUBLIC_COLUMNS = [
    "public_id", "title", "description", "property_type", "room_type", "status",
    "city", "state", "country", "latitude", "longitude",
    "price_per_night", "cleaning_fee", "max_guests", "bedrooms", "beds", "bathrooms",
    "rating_average", "rating_count", "created_at",
]
# `address_line1` and `postal_code` are deliberately absent above: Airbnb reveals the exact
# address only after booking, and publishing it says which houses are empty next week.
PROPERTY_OWNER_COLUMNS = PROPERTY_PUBLIC_COLUMNS + ["address_line1", "postal_code"]

BOOKING_COLUMNS = [
    "public_id", "check_in", "check_out", "guests", "nights",
    "nightly_rate", "subtotal", "cleaning_fee", "service_fee", "total",
    "status", "cancelled_at", "cancellation_reason", "created_at",
]

# --- row-level filters ------------------------------------------------------------

PUBLISHED_ONLY = {"_and": [{"status": {"_eq": "PUBLISHED"}}, {"deleted": {"_eq": False}}]}
NOTHING_HIDDEN = {}  # an empty filter means "every row"

# ⚠️ "Staff see everything" does NOT mean an empty filter on a soft-deleting table. A deleted row
# is gone as far as the product is concerned, and `{}` shows it — so the admin console counted
# listings that had been removed while `/api/v1/admin/stats`, which filters `deleted = false` in
# SQL, did not. Two totals for the same thing, both plausible, and only one right.
#
# This is the Hasura shape of the classic soft-delete trap: the ORM applies the flag automatically,
# and anything written by hand — SQL or a permission rule — has to say so itself.
NOT_DELETED = {"deleted": {"_eq": False}}

# "the property whose host is me" — a filter that walks the relationship rather than duplicating
# host_id onto every child table.
OWN_PROPERTY = {"host": {"public_id": {"_eq": f"{SESSION_USER}"}}}
OWN_BOOKING = {"guest": {"public_id": {"_eq": f"{SESSION_USER}"}}}
BOOKING_AT_OWN_PROPERTY = {"property": OWN_PROPERTY}


def _select(columns: list[str], filter_: dict, *, limit: int | None = None) -> dict:
    perm = {"columns": columns, "filter": filter_, "allow_aggregations": True}
    if limit is not None:
        # A hard cap Hasura enforces server-side. Without it, one `query { properties }` with no
        # arguments returns the entire table — the classic way a GraphQL endpoint becomes a
        # denial-of-service vector against its own database.
        perm["limit"] = limit
    return perm


def build() -> dict:
    """The complete metadata document, ready for `replace_metadata`."""
    return {
        "version": 3,
        "sources": [
            {
                "name": SOURCE,
                "kind": "postgres",
                "configuration": {
                    "connection_info": {
                        "database_url": {"from_env": "HASURA_GRAPHQL_DATABASE_URL"},
                        "isolation_level": "read-committed",
                        "use_prepared_statements": True,
                    }
                },
                # Matches HASURA_GRAPHQL_DEFAULT_NAMING_CONVENTION in docker-compose.yml, so
                # columns are exposed as `pricePerNight` rather than `price_per_night` — the same
                # shape FastAPI returns.
                "customization": {"naming_convention": "graphql-default"},
                "tables": _tables(),
            }
        ],
    }


def _tables() -> list[dict]:
    return [
        # ---------------------------------------------------------------- users
        {
            "table": {"schema": "public", "name": "users"},
            "array_relationships": [
                {
                    "name": "properties",
                    "using": {
                        "foreign_key_constraint_on": {
                            "table": {"schema": "public", "name": "properties"},
                            "column": "host_id",
                        }
                    },
                },
                {
                    "name": "bookings",
                    "using": {
                        "foreign_key_constraint_on": {
                            "table": {"schema": "public", "name": "bookings"},
                            "column": "guest_id",
                        }
                    },
                },
            ],
            "select_permissions": [
                {
                    "role": "anonymous",
                    # A visitor may read a HOST's public profile — the name and bio shown on a
                    # listing — and nobody else's. Without the is_host filter this exposes every
                    # registered user's name to the open internet.
                    "permission": _select(
                        USER_PUBLIC_COLUMNS,
                        {"_and": [{"is_host": {"_eq": True}}, {"deleted": {"_eq": False}}]},
                        limit=50,
                    ),
                },
                {
                    "role": "customer",
                    "permission": _select(
                        USER_PUBLIC_COLUMNS,
                        {"_and": [{"is_host": {"_eq": True}}, {"deleted": {"_eq": False}}]},
                        limit=50,
                    ),
                },
                {
                    "role": "host",
                    # A host needs to know WHO booked their place — otherwise the reservations
                    # page shows a stay with a blank guest, and the relationship silently
                    # resolves to null rather than erroring, which looks like a broken join.
                    #
                    # The second branch walks guest → booking → property → host. Reading a
                    # guest's profile is therefore earned by them having booked with you; it is
                    # not a blanket "hosts can see all users".
                    "permission": _select(
                        USER_PUBLIC_COLUMNS,
                        {
                            "_or": [
                                {"_and": [{"is_host": {"_eq": True}}, {"deleted": {"_eq": False}}]},
                                {"bookings": BOOKING_AT_OWN_PROPERTY},
                            ]
                        },
                        limit=50,
                    ),
                },
                {"role": "staff", "permission": _select(USER_SELF_COLUMNS, NOT_DELETED)},
            ],
        },
        # ------------------------------------------------------------ properties
        {
            "table": {"schema": "public", "name": "properties"},
            "object_relationships": [
                {"name": "host", "using": {"foreign_key_constraint_on": "host_id"}},
            ],
            "array_relationships": [
                {
                    "name": "images",
                    "using": {
                        "foreign_key_constraint_on": {
                            "table": {"schema": "public", "name": "property_images"},
                            "column": "property_id",
                        }
                    },
                },
                {
                    # ⚠️ camelCase by hand. `graphql-default` renames COLUMNS, not relationship
                    # names — those are taken verbatim from metadata. The giveaway is that its
                    # derived aggregate field IS camelCased, so a snake_case relationship called
                    # `property_amenities` sits next to `propertyAmenitiesAggregate` in the same
                    # type, and a query written to match the rest of the schema fails with
                    # "field 'propertyAmenities' not found in type: 'Properties'".
                    "name": "propertyAmenities",
                    "using": {
                        "foreign_key_constraint_on": {
                            "table": {"schema": "public", "name": "property_amenities"},
                            "column": "property_id",
                        }
                    },
                },
                {
                    "name": "bookings",
                    "using": {
                        "foreign_key_constraint_on": {
                            "table": {"schema": "public", "name": "bookings"},
                            "column": "property_id",
                        }
                    },
                },
            ],
            "select_permissions": [
                {"role": "anonymous", "permission": _select(PROPERTY_PUBLIC_COLUMNS, PUBLISHED_ONLY, limit=100)},
                {"role": "customer", "permission": _select(PROPERTY_PUBLIC_COLUMNS, PUBLISHED_ONLY, limit=100)},
                {
                    "role": "host",
                    # A host sees published listings PLUS their own drafts. `_or`, not a
                    # replacement — a host is still a guest browsing the site.
                    "permission": _select(
                        PROPERTY_OWNER_COLUMNS,
                        {"_or": [PUBLISHED_ONLY, {"_and": [OWN_PROPERTY, {"deleted": {"_eq": False}}]}]},
                        limit=100,
                    ),
                },
                {"role": "staff", "permission": _select(PROPERTY_OWNER_COLUMNS, NOT_DELETED)},
            ],
        },
        # -------------------------------------------------------- property_images
        {
            "table": {"schema": "public", "name": "property_images"},
            "object_relationships": [
                {"name": "property", "using": {"foreign_key_constraint_on": "property_id"}},
            ],
            "select_permissions": [
                # ⚠️ Permissions do NOT cascade through relationships. Being allowed to read a
                # property does not make its images readable — each table needs its own rule, and
                # each rule re-states the visibility condition by walking back to the parent.
                # Miss this and a listing page renders with no photos and no error.
                {
                    "role": r,
                    "permission": _select(
                        ["url", "alt_text", "sort_order", "is_cover"],
                        {"property": PUBLISHED_ONLY} if r in ("anonymous", "customer")
                        else {"_or": [{"property": PUBLISHED_ONLY}, {"property": OWN_PROPERTY}]},
                        limit=200,
                    ),
                }
                for r in ("anonymous", "customer", "host")
            ]
            + [
                {
                    "role": "staff",
                    "permission": _select(
                        ["url", "alt_text", "sort_order", "is_cover"],
                        {"property": NOT_DELETED},
                    ),
                }
            ],
        },
        # ------------------------------------------------------------- amenities
        {
            "table": {"schema": "public", "name": "amenities"},
            "select_permissions": [
                # A static vocabulary. Public to everyone, with no row filter, because there is
                # nothing here worth hiding.
                {"role": r, "permission": _select(["slug", "name", "icon"], NOTHING_HIDDEN)}
                for r in ("anonymous", "customer", "host", "staff")
            ],
        },
        # ---------------------------------------------------- property_amenities
        {
            "table": {"schema": "public", "name": "property_amenities"},
            "object_relationships": [
                {"name": "property", "using": {"foreign_key_constraint_on": "property_id"}},
                {"name": "amenity", "using": {"foreign_key_constraint_on": "amenity_id"}},
            ],
            "select_permissions": [
                {"role": r, "permission": _select([], NOTHING_HIDDEN, limit=500)}
                for r in ("anonymous", "customer", "host", "staff")
            ],
        },
        # -------------------------------------------------------------- bookings
        {
            "table": {"schema": "public", "name": "bookings"},
            "object_relationships": [
                {"name": "property", "using": {"foreign_key_constraint_on": "property_id"}},
                {"name": "guest", "using": {"foreign_key_constraint_on": "guest_id"}},
            ],
            "select_permissions": [
                # No anonymous rule at all. An absent permission means the table is INVISIBLE to
                # that role — it does not even appear in the schema — which is stronger and
                # clearer than a rule that filters everything out.
                {"role": "customer", "permission": _select(BOOKING_COLUMNS, OWN_BOOKING)},
                {
                    "role": "host",
                    # Their own stays, plus reservations at their listings. A host reading a
                    # booking at their own place is the /hosts/reservations page.
                    "permission": _select(
                        BOOKING_COLUMNS, {"_or": [OWN_BOOKING, BOOKING_AT_OWN_PROPERTY]}
                    ),
                },
                {"role": "staff", "permission": _select(BOOKING_COLUMNS, NOTHING_HIDDEN)},
            ],
        },
        # -------------------------------------------------------------- payments
        {
            "table": {"schema": "public", "name": "payments"},
            "object_relationships": [
                {"name": "booking", "using": {"foreign_key_constraint_on": "booking_id"}},
            ],
            "select_permissions": [
                {
                    "role": "customer",
                    "permission": _select(
                        ["public_id", "status", "amount", "currency", "card_brand", "card_last4", "created_at"],
                        {"booking": OWN_BOOKING},
                    ),
                },
                # Deliberately NOT visible to hosts. A host is entitled to know a booking is paid
                # — which `bookings.status` already tells them — not to see the guest's card
                # metadata.
                {
                    "role": "staff",
                    "permission": _select(
                        ["public_id", "status", "amount", "currency", "card_brand", "card_last4",
                         "failure_message", "stripe_payment_intent_id", "created_at"],
                        NOTHING_HIDDEN,
                    ),
                },
            ],
        },
        # --------------------------------------------------------------- reviews
        {
            "table": {"schema": "public", "name": "reviews"},
            "object_relationships": [
                {"name": "property", "using": {"foreign_key_constraint_on": "property_id"}},
                {"name": "author", "using": {"foreign_key_constraint_on": "author_id"}},
            ],
            "select_permissions": [
                {
                    "role": r,
                    "permission": _select(
                        ["public_id", "rating", "comment", "created_at"],
                        {"property": PUBLISHED_ONLY},
                        limit=100,
                    ),
                }
                for r in ("anonymous", "customer", "host")
            ]
            + [
                {
                    "role": "staff",
                    "permission": _select(["public_id", "rating", "comment", "created_at"], NOTHING_HIDDEN),
                }
            ],
        },
    ]

File: contractor/hasura/metadata.mjs
/**
 * The Hasura metadata: which tables exist, how they relate, and who may read what.
 *
 * This is the whole read side of the app. NestJS owns every write; Hasura owns every read, and the
 * permissions below are what make that safe. Written as JavaScript rather than the YAML the Hasura
 * CLI produces, for one reason: the interesting part of a permission is WHY a filter is shaped the
 * way it is, and YAML has nowhere to put that.
 *
 * ══════════════════════════════════════════════════════════════════════════════════════════════
 * ⚠️ NO ROLE HAS AN INSERT, UPDATE OR DELETE PERMISSION. ANYWHERE. THAT IS THE DESIGN.
 * ══════════════════════════════════════════════════════════════════════════════════════════════
 *
 * Hasura can generate mutations for every table, and it is genuinely tempting — one line of
 * metadata and the frontend can insert a quote with no backend code at all. It also means:
 *
 *   • accepting a quote no longer declines the others, because that is three writes in a
 *     transaction and a generated mutation is one write;
 *   • a contractor can quote on a trade they do not work in, because "may I see this project"
 *     and "may I bid on it" are different questions and a select permission only answers the first;
 *   • a homeowner can review a job that was never completed;
 *   • a contractor can set their own rating.
 *
 * Every one of those rules lives in a NestJS service. Adding a single insert permission here opens
 * a second path to the same tables that does not go past any of them.
 *
 * Four roles:
 *   anonymous   — not signed in. Browses the public directory and nothing else.
 *   homeowner   — their own projects, and the quotes on them.
 *   contractor  — leads in their trades, and THEIR OWN quotes. Never a competitor's.
 *   staff       — everything, for support.
 *
 * ⚠️ The staff role is called `staff` and NOT `admin`. `admin` is Hasura's own built-in superuser
 * role: it bypasses every permission in this file, and Hasura refuses to let you define
 * permissions for it ("cannot define permission for admin role"). A JWT claiming it would get
 * unrestricted read AND WRITE access to every table — which is exactly the door this file exists
 * to keep shut. See `common/enums.ts` in the backend.
 */

const PUBLIC = ['anonymous', 'homeowner', 'contractor', 'staff']

/** The session variables the JWT carries. See `auth/jwt-payload.ts` — both are strings. */
const USER_ID = 'X-Hasura-User-Id'
const CONTRACTOR_ID = 'X-Hasura-Contractor-Id'

/** Builds the same select permission for several roles at once. */
function selectFor(roles, permission) {
  return roles.map((role) => ({ role, permission }))
}

/**
 * ⚠️ Column allowlists, never `columns: '*'`.
 *
 * A wildcard means "every column this table has NOW and every column anyone adds later". The day
 * somebody adds `internal_notes` to `projects`, a wildcard hands it to every homeowner on the site
 * and nothing in the diff looks like a permission change. Listing columns makes exposing a new one
 * a deliberate act.
 */
const USER_PUBLIC_COLUMNS = ['public_id', 'first_name', 'last_name', 'avatar_url', 'role', 'created_at']

const CONTRACTOR_COLUMNS = [
  'public_id',
  'user_id',
  'business_name',
  'bio',
  'years_in_business',
  'license_number',
  'city',
  'state',
  'zip',
  'service_radius_miles',
  'hourly_rate_min',
  'hourly_rate_max',
  'rating_average',
  'review_count',
  'created_at',
]

const PROJECT_COLUMNS = [
  'public_id',
  'homeowner_id',
  'service_category_id',
  'title',
  'description',
  'city',
  'state',
  'zip',
  'budget_min',
  'budget_max',
  'preferred_start_date',
  'status',
  'created_at',
]

const QUOTE_COLUMNS = [
  'public_id',
  'project_id',
  'contractor_profile_id',
  'amount',
  'estimated_days',
  'message',
  'status',
  'created_at',
]

const REVIEW_COLUMNS = [
  'public_id',
  'project_id',
  // ⚠️ Denormalised onto the review row on purpose. `projects` is NOT readable by `anonymous`,
  // and Hasura hides a relationship field pointing at a table the role cannot select — so
  // `review.project.title` on a public profile fails with "field 'project' not found in type:
  // 'reviews'". See the AddProjectTitleToReviews migration.
  'project_title',
  'homeowner_id',
  'contractor_profile_id',
  'rating',
  'comment',
  'created_at',
]

/**
 * "A project in a trade this contractor works in."
 *
 * Reads across the join table: project → its category → the rows linking that category to a
 * profile → is one of them mine. This is rule 1 expressed as a permission, and it is the same rule
 * `QuotesService` checks before accepting a bid. Both are needed: this one stops the pro from
 * SEEING the job, the service one stops them from quoting on an id they got some other way.
 */
const IN_MY_TRADES = {
  category: { contractor_services: { contractor_profile_id: { _eq: CONTRACTOR_ID } } },
}

export const metadata = {
  version: 3,
  sources: [
    {
      name: 'default',
      kind: 'postgres',
      configuration: {
        connection_info: {
          /**
           * ⚠️ `from_env`, not a literal connection string. `replace_metadata` overwrites the
           * whole source definition, so a literal here would bake the password into a committed
           * file AND into Hasura's own metadata database. This just points at the variable
           * docker-compose.yml already sets.
           */
          database_url: { from_env: 'HASURA_GRAPHQL_DATABASE_URL' },
          isolation_level: 'read-committed',
          use_prepared_statements: true,
        },
      },

      tables: [
        /* ------------------------------------------------------------------------------------ */
        /* users — deliberately the most restricted table here                                    */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'users' },
          object_relationships: [
            {
              name: 'contractor_profile',
              using: { manual_configuration: {
                remote_table: { schema: 'public', name: 'contractor_profiles' },
                column_mapping: { id: 'user_id' },
              } },
            },
          ],
          array_relationships: [
            {
              name: 'projects',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'projects' },
                column: 'homeowner_id',
              } },
            },
          ],
          /**
           * ⚠️ `email`, `phone` and above all `password_hash` are absent from this list, for every
           * role including admin. A homeowner's own email reaches them through `/auth/login`, not
           * through GraphQL — so there is no query anywhere in this app that can return one, and
           * therefore no permission to get subtly wrong later.
           */
          select_permissions: selectFor(PUBLIC, {
            columns: USER_PUBLIC_COLUMNS,
            // Soft-deleted accounts vanish from every read. The rows stay, because reviews and
            // quotes reference them forever.
            filter: { deleted: { _eq: false } },
          }),
        },

        /* ------------------------------------------------------------------------------------ */
        /* service_categories — fully public, read by the home page before anyone signs in        */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'service_categories' },
          array_relationships: [
            {
              name: 'contractor_services',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'contractor_services' },
                column: 'service_category_id',
              } },
            },
            {
              name: 'projects',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'projects' },
                column: 'service_category_id',
              } },
            },
          ],
          select_permissions: selectFor(PUBLIC, {
            columns: ['public_id', 'slug', 'name', 'description', 'icon'],
            filter: {},
            allow_aggregations: true,
          }),
        },

        /* ------------------------------------------------------------------------------------ */
        /* contractor_profiles — the public directory                                             */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'contractor_profiles' },
          object_relationships: [
            { name: 'user', using: { foreign_key_constraint_on: 'user_id' } },
          ],
          array_relationships: [
            {
              name: 'contractor_services',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'contractor_services' },
                column: 'contractor_profile_id',
              } },
            },
            {
              name: 'portfolio_images',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'portfolio_images' },
                column: 'contractor_profile_id',
              } },
            },
            {
              name: 'reviews',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'reviews' },
                column: 'contractor_profile_id',
              } },
            },
            {
              name: 'quotes',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'quotes' },
                column: 'contractor_profile_id',
              } },
            },
          ],
          select_permissions: selectFor(PUBLIC, {
            columns: CONTRACTOR_COLUMNS,
            filter: { deleted: { _eq: false } },
            // The directory sorts by rating and shows counts, both of which are aggregates.
            allow_aggregations: true,
          }),
        },

        /* ------------------------------------------------------------------------------------ */
        /* contractor_services — the join table                                                   */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'contractor_services' },
          object_relationships: [
            { name: 'contractor_profile', using: { foreign_key_constraint_on: 'contractor_profile_id' } },
            { name: 'category', using: { foreign_key_constraint_on: 'service_category_id' } },
          ],
          /**
           * ⚠️ Readable by everyone, and it has to be — the `IN_MY_TRADES` filter on `projects`
           * traverses this table. Hasura evaluates a permission's relationship hops against the
           * TABLE's permissions too, so a locked-down join table makes the filter that depends on
           * it match nothing, silently. It holds two foreign keys and no secrets.
           */
          select_permissions: selectFor(PUBLIC, {
            columns: ['contractor_profile_id', 'service_category_id'],
            filter: {},
          }),
        },

        /* ------------------------------------------------------------------------------------ */
        /* portfolio_images — public, they are the point of the profile page                      */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'portfolio_images' },
          object_relationships: [
            { name: 'contractor_profile', using: { foreign_key_constraint_on: 'contractor_profile_id' } },
          ],
          select_permissions: selectFor(PUBLIC, {
            columns: ['public_id', 'contractor_profile_id', 'url', 'caption', 'sort_order', 'created_at'],
            filter: {},
          }),
        },

        /* ------------------------------------------------------------------------------------ */
        /* projects — where the roles genuinely diverge                                           */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'projects' },
          object_relationships: [
            { name: 'homeowner', using: { foreign_key_constraint_on: 'homeowner_id' } },
            { name: 'category', using: { foreign_key_constraint_on: 'service_category_id' } },
            {
              name: 'review',
              using: { manual_configuration: {
                remote_table: { schema: 'public', name: 'reviews' },
                column_mapping: { id: 'project_id' },
              } },
            },
          ],
          array_relationships: [
            {
              name: 'quotes',
              using: { foreign_key_constraint_on: {
                table: { schema: 'public', name: 'quotes' },
                column: 'project_id',
              } },
            },
          ],
          select_permissions: [
            /**
             * ⚠️ `anonymous` has NO entry here, and an absent permission is a deny. Someone's
             * home address, budget and schedule are not public, and this is the table where
             * forgetting that would matter most.
             */
            {
              role: 'homeowner',
              permission: {
                columns: PROJECT_COLUMNS,
                // Their own, and only their own. `X-Hasura-User-Id` comes from a signature Hasura
                // verified — it cannot be set by the caller.
                filter: { homeowner_id: { _eq: USER_ID } },
                allow_aggregations: true,
              },
            },
            {
              role: 'contractor',
              permission: {
                columns: PROJECT_COLUMNS,
                /**
                 * Two ways a pro may see a project, and both are needed:
                 *
                 *  1. it is OPEN and in one of their trades — the lead feed;
                 *  2. they have already quoted on it — "My quotes", which must keep working after
                 *     the project is hired, in progress or completed, at which point rule 1 no
                 *     longer matches it.
                 *
                 * Without the second clause a contractor loses sight of a job the moment they win
                 * it, which is precisely when they need it.
                 */
                filter: {
                  _or: [
                    { _and: [{ status: { _in: ['open', 'quoted'] } }, IN_MY_TRADES] },
                    { quotes: { contractor_profile_id: { _eq: CONTRACTOR_ID } } },
                  ],
                },
                allow_aggregations: true,
              },
            },
            {
              role: 'staff',
              permission: { columns: PROJECT_COLUMNS, filter: {}, allow_aggregations: true },
            },
          ],
        },

        /* ------------------------------------------------------------------------------------ */
        /* quotes — the table where a wrong filter would break the marketplace                    */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'quotes' },
          object_relationships: [
            { name: 'project', using: { foreign_key_constraint_on: 'project_id' } },
            { name: 'contractor_profile', using: { foreign_key_constraint_on: 'contractor_profile_id' } },
          ],
          select_permissions: [
            {
              role: 'homeowner',
              permission: {
                columns: QUOTE_COLUMNS,
                // Every quote on a project they own — that is the comparison screen.
                filter: { project: { homeowner_id: { _eq: USER_ID } } },
                allow_aggregations: true,
              },
            },
            {
              role: 'contractor',
              permission: {
                columns: QUOTE_COLUMNS,
                /**
                 * ⚠️ THEIR OWN QUOTES ONLY. Never a competitor's.
                 *
                 * The obvious-looking alternative — "quotes on projects I can see" — reads
                 * naturally and destroys the product: the last pro to bid could read every rival
                 * price and undercut by a dollar, and after that nobody quotes honestly. This is
                 * the single most consequential line in this file.
                 */
                filter: { contractor_profile_id: { _eq: CONTRACTOR_ID } },
                allow_aggregations: true,
              },
            },
            {
              role: 'staff',
              permission: { columns: QUOTE_COLUMNS, filter: {}, allow_aggregations: true },
            },
          ],
        },

        /* ------------------------------------------------------------------------------------ */
        /* reviews — public, because they are the reason the directory is worth reading           */
        /* ------------------------------------------------------------------------------------ */
        {
          table: { schema: 'public', name: 'reviews' },
          object_relationships: [
            { name: 'project', using: { foreign_key_constraint_on: 'project_id' } },
            { name: 'homeowner', using: { foreign_key_constraint_on: 'homeowner_id' } },
            { name: 'contractor_profile', using: { foreign_key_constraint_on: 'contractor_profile_id' } },
          ],
          select_permissions: selectFor(PUBLIC, {
            columns: REVIEW_COLUMNS,
            filter: {},
            allow_aggregations: true,
          }),
        },
      ],
    },
  ],
}
